sys/fork.c

```diff
@@ -0,0 +1,55 @@
+#include <fork.h>
+#include <thread.h>
+#include <sync.h>
+#include <filedesc.h>
+#include <sched.h>
+#include <stdc.h>
+#include <vm_map.h>
+
+int do_fork() {
+  thread_t *td = thread_self();
+
+  /* Cannot fork non-user threads. */
+  assert(td->td_uspace);
+
+  thread_t *newtd = thread_create(td->td_name, NULL, NULL);
+
+  /* Clone the thread. Since we don't use fork-oriented thread_t layout, we copy
+     all fields one-by one for clarity. The new thread is already on the
+     all_thread list, has name and tid set. */
+
+  newtd->td_state = td->td_state;
+  newtd->td_flags = td->td_flags;
```

Comment:
I see no reason to copy it from the parent. It will be probably zeroed out by `sched_add`.